mono/utils/mono-os-wait-win32.c

```diff
@@ -121,29 +123,48 @@ mono_win32_wait_for_single_object_ex (HANDLE handle, DWORD timeout, BOOL alertab
 
 	// NOTE, leave_alertable_wait should not affect GetLastError but
 	// if changed, GetLastError needs to be preserved and reset before returning.
-	if (alertable && info) {
+	if (info)
 		leave_alertable_wait (info);
-	}
 
 	return result;
 }
 
 DWORD
-mono_win32_wait_for_multiple_objects_ex (DWORD count, CONST HANDLE *handles, BOOL waitAll, DWORD timeout, BOOL alertable)
+mono_win32_wait_for_multiple_objects_ex (DWORD count, CONST HANDLE *handles, BOOL waitAll, DWORD timeout, BOOL alertable, MonoError *error)
 {
-	DWORD result = WAIT_FAILED;
 	MonoThreadInfo * const info = alertable ? mono_thread_info_current_unchecked () : NULL;
 
 	if (info)
 		enter_alertable_wait (info);
 
-	result = WaitForMultipleObjectsEx (count, handles, waitAll, timeout, alertable);
+	DWORD const result = WaitForMultipleObjectsEx (count, handles, waitAll, timeout, alertable);
 
 	// NOTE, leave_alertable_wait should not affect GetLastError but
 	// if changed, GetLastError needs to be preserved and reset before returning.
 	if (info)
 		leave_alertable_wait (info);
 
+	// This is not perfect, but it is the best you can do in usermode and matches CoreCLR.
+	// i.e. handle-based instead of object-based.
+
+	if (result == WAIT_FAILED && waitAll && error &&
+			count > 1 && count <= MAXIMUM_WAIT_OBJECTS
+			&& GetLastError () == ERROR_INVALID_PARAMETER) {
+		gpointer handles_sorted [MAXIMUM_WAIT_OBJECTS]; // 64
+		memcpy (handles_sorted, handles, count * sizeof (handles [0]));
+		qsort (handles_sorted, count, sizeof (gpointer), g_direct_equal);
```

Comment:
Just for clarity and consistency, `qsort (handles_sorted, count, sizeof  (handles_sorted [0]), g_direct_equal);`